File: data-structures/linkedlist/linkedlist.cc
#include <iostream>
#include <stdlib.h>

struct Node{
  int value;
  struct Node *next;
};

struct Linkedlist{
  struct Node *head;
};

struct Linkedlist* createLinkedlist(){
  struct Linkedlist* list = (struct Linkedlist*)(malloc(sizeof(struct Linkedlist)));
  list->head = NULL;
  return list;
}

void addNode(struct Linkedlist* list, int element){
  struct Node* temp = (struct Node*)malloc(sizeof(struct Node));
  temp->value = element;
  temp->next = list->head;
  list->head = temp;
}

void deleteNode(struct Linkedlist* list, int element){
  struct Node* currentNode = list->head;
  struct Node* prev = NULL;
  int found = 0;

  while (found == 0){
    if (currentNode->value==element){
      found = 1;
    }
    else{
      prev = currentNode;
      currentNode = currentNode->next;
    }
  }

  if(prev == NULL){
    list->head = currentNode->next;
  }
  else{
    prev->next = currentNode->next;
  }
}

int containsNode(struct Linkedlist* list, int element){
  struct Node* currentNode = list->head;
  int found = 0;

  while(found == 0){
    if(currentNode->value == element){
      found = 1;
      return found;
    }
    else{
      currentNode = currentNode->next;
    }
  }

  return found;
}

void display(struct Linkedlist* list){
  struct Node* currentNode = list->head;

  while(currentNode != NULL){
    std::cout << currentNode->value << std::endl;
    currentNode = currentNode->next;
  }
}

int main(){
  struct Linkedlist* list = createLinkedlist();
  addNode(list, 23);
  addNode(list, 4);
  display(list);
  deleteNode(list, 4);
  display(list);
  return 0;
}

File: data-structures/linkedlist/linkedlist.py
class Node:

	def __init__(self, value=None, next=None):
		self.value = value
		self.next = next

class Linkedlist:

	def __init__(self, head=None):
		self.head = head

	def add_node(self, value):
		new_node = Node(value)
		new_node.next = self.head
		self.head = new_node

	def delete_node(self, value):
		current_node = self.head
		prev = None
		found = False

		while found is False:
			if current_node.value == value:
				found = True
			else:
				prev = current_node
				current_node = current_node.next

		if prev == None:
			self.head = current_node.next
		else:
			prev.next = current_node.next

	def length(self):
		current_node = self.head
		count = 0
		while current_node != None:
			count += 1
			current_node = current_node.next

		return count

	def get_position(self, value):
		count = 1
		found = False
		current_node = self.head

		while not found:
			if current_node.value == value:
				found = True
				print count
			else:
				current_node = current_node.next
				count += 1

	def display(self):
		current_node = self.head

		while current_node != None:
			print current_node.value
			current_node = current_node.next

	def is_palindrome(self):
		#2.7
		current_node = self.head
		word = ""

		while current_node != None:
			word += current_node.value
			current_node = current_node.next

		if word == word[::-1]:
			return True

		return False

	def delete_middle_node(self):
		#2.4
		current_node = self.head
		l = self.length()
		mid = l/2
		prev = None
		count = 0
		found = False

		while found is False:
			if count == mid:
				found = True
			else:
				prev = current_node
				current_node = current_node.next
				count += 1

		if prev == None:
			self.head = current_node.next
		else:
			prev.next = current_node.next

#testing

test = Linkedlist()
test.add_node('j')
test.add_node('b') #points to 30
test.add_node('o') #points to 29
test.add_node('b') #points to 25 [16, 25, 29, 30]
test.add_node('j')

# print test.is_palindrome()

test.delete_middle_node()
test.display()

# test.delete_node(29)
